### PR TITLE
Refactor Vote Collection

### PR DESCRIPTION
**Pull request by bfish713**
We collect votes in 3 places and the logic was just duplicated everywhere.  This removes most of that duplication and puts it into an `Accumlator`.  There is still some funkiness in election because it has both DA and QC certificate types.  I also couldn't figure a way not have a `accumulate_da_vote` and `accumulate_qc_vote` functions despite the logic looking the same

Closes: https://github.com/EspressoSystems/HotShot/issues/883

### CODE REVIEW

File: types/src/traits/election.rs [hunk 1]
```diff
@@ -101,9 +102,33 @@ pub trait Accumulator<T, U>: Sized {
...
+//         valid_signatures.insert(key, (sig.clone(), token.clone()));
+
+//         stake_casted += u64::from(token.vote_count());
+
+//         if stake_casted >= u64::from(threshold) {
+//             return Either::Right((self.valid_signatures));
+//         }
+//         Either::Left(self)
+//     }
+// }
```
Comment (DieracDelta) [inline types/src/traits/election.rs:131]: nit: Can this be deleted?

File: types/src/traits/election.rs [hunk 2]
```diff
@@ -113,7 +138,17 @@ where
     Self: Send + Sync + Clone + Serialize + for<'a> Deserialize<'a>,
     LEAF: Committable,
 {
-    type Accumulator: Accumulator<(EncodedSignature, SIGNATURE), Self>;
+    // type Accumulator: Accumulator<
+    //     (Commitment<LEAF>, (EncodedPublicKey, (EncodedSignature, TOKEN))),
+    //     BTreeMap<EncodedPublicKey, (EncodedSignature, TOKEN)>,
+    // >;
```
Comment (DieracDelta) [inline types/src/traits/election.rs:144]: nit: Can this be deleted?

File: types/src/traits/election.rs [hunk 3]
```diff
@@ -234,6 +269,83 @@ pub trait Election<TYPES: NodeType>: Clone + Eq + PartialEq + Send + Sync + 'sta
...
+    ) -> Either<
+        CertificateAccumulator<TYPES::VoteTokenType, Self::LeafType>,
+        QuorumCertificate<TYPES, Self::LeafType>,
+    > {
+        if !self.is_valid_vote(
+            encoded_key,
+            encoded_signature,
+            VoteData::Yes(leaf_commitment),
+            view_number,
+            // Ignoring deserialization errors below since we are getting rid of it soon
```
Comment (DieracDelta) [inline types/src/traits/election.rs:291]: I think this was copied over from before, but I'm not sure what exactly we're getting rid of. Serialiazation? Does anyone know? If not we can probably delete the comment and the one below

File: types/src/traits/election.rs [hunk 4]
```diff
@@ -234,6 +269,83 @@ pub trait Election<TYPES: NodeType>: Clone + Eq + PartialEq + Send + Sync + 'sta
...
+    /// Accumlate the vote, return the QC if the threshold was reached, if not return the updated accumlator
+    #[allow(clippy::type_complexity)]
+    fn accumulate_qc_vote(
+        &self,
+        encoded_key: &EncodedPublicKey,
+        encoded_signature: &EncodedSignature,
+        leaf_commitment: Commitment<Self::LeafType>,
+        vote_token: TYPES::VoteTokenType,
+        view_number: TYPES::Time,
+        accumlator: CertificateAccumulator<TYPES::VoteTokenType, Self::LeafType>,
```
Comment (DieracDelta) [inline types/src/traits/election.rs:281]: nit: `accumulator`

File: types/src/traits/election.rs [hunk 5]
```diff
@@ -234,6 +269,83 @@ pub trait Election<TYPES: NodeType>: Clone + Eq + PartialEq + Send + Sync + 'sta
         is_valid_signature && is_valid_vote_token
     }
 
+    /// Accumlate the vote, return the QC if the threshold was reached, if not return the updated accumlator
+    #[allow(clippy::type_complexity)]
+    fn accumulate_qc_vote(
+        &self,
+        encoded_key: &EncodedPublicKey,
+        encoded_signature: &EncodedSignature,
+        leaf_commitment: Commitment<Self::LeafType>,
```
Comment (DieracDelta) [inline types/src/traits/election.rs:278]: An idea to reduce code duplication:

Maybe we could stick all these arguments into a struct, and make that struct generic over the type of commitment? This will fix the clippy lint. Denote this struct `VoteMetaData<T>`. 

We could then type alias `pub type QcVote<WhateverSelfIs> = VoteMetaData<WhateverSelfIs::LeafType` and `pub type DACVote<WhateverSelfIs> = VoteMetaData<WhateverSelfIs::Block>`. Then this `accumulate_vote` could be made generic over the commitment type and certificate type, and we could have const function aliases as part of the trait that monomorphize over the struct type aliases and certificate types.

**Review by DieracDelta — DISMISSED** [review]
LGTM! I had one thought on merging those two functions

File: src/traits/election/static_committee.rs [hunk 6]
```diff
@@ -95,7 +95,7 @@ where
 
     type QuorumCertificate = LEAF::QuorumCertificate;
 
-    type DACertificate = LEAF::DACertificate;
+    type DACertificate = DACertificate<TYPES>;
```
Comment (shenkeyao) [inline src/traits/election/static_committee.rs:98]: Nit: I'm curious why this change is needed for DAC but not for QC.

File: types/src/traits/election.rs (hunk 3 above)
Comment (shenkeyao) [reply types/src/traits/election.rs:291]: Maybe the code below this comment used to do some deserialization, and we forgot to remove/update the comment? It doesn't seem to be related to serialization now. I guess we can delete it.

**Review by shenkeyao — APPROVED** [review]
LGTM

File: src/traits/election/static_committee.rs (hunk 6 above)
Comment (bfish713) [reply src/traits/election/static_committee.rs:98]: It's because the DAC is based of the blocks commitment rather than the leaf commitment and the LeafType::DAC bounds DAC with `SignedCertificate<..., Self>`.  So the bounds are mismatched with Election bound DAC with `SignedCertificate<..., TYPES::BlockType>`

I actually reverted this and updated the trait bounds in LeafType and that seems to work and make more sense

File: types/src/traits/election.rs (hunk 5 above)
Comment (bfish713) [reply types/src/traits/election.rs]: Based of this suggestion I did something similar to this but wasn't sure about the type alias part.  LMK what you think

Comment (DieracDelta) [reply types/src/traits/election.rs]: The type aliases were meant as a convenience so we don't accidentally feed in the wrong commitment. But, it looks like Rust is inferring the type where needed, which is cleaner IMO. Nice, LGTM!

**Review by DieracDelta — APPROVED** [review]

**Review by shenkeyao — APPROVED** [review]
